### PR TITLE
#4: Tunnelling

### LINKED ISSUE #4: Tunnelling
1. Switch off for all
2. Switch on for required clients
```
ip rule add from $ip table $table
ip route add default dev wg table $table
```

### PR DESCRIPTION
**Pull request by KirillSmirnow**

### CODE REVIEW

File: vpn-router-core/src/main/java/vpnrouter/core/infrastructure/shell/RealTunnelling.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+import static java.util.stream.Collectors.joining;
+
+@Primary
+@Component
+@Profile("prod")
+@RequiredArgsConstructor
+public class RealTunnelling implements Tunnelling {
+
+    private static final int TABLE = 7373;
```
Comment (shurshurik) [inline vpn-router-core/src/main/java/vpnrouter/core/infrastructure/shell/RealTunnelling.java:19]: why is this value set for TABLE?

File: vpn-router-core/src/main/java/vpnrouter/core/infrastructure/shell/RealTunnelling.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    private final NetworkProperties networkProperties;
+    private final ShellExecutor shellExecutor;
+
+    @Override
+    public void switchOnOnlyFor(Set<String> ipAddresses) {
+        shellExecutor.execute(
+                buildResetTableCommand() + buildSwitchOffCommand() + buildSwitchOnCommand(ipAddresses)
```
Comment (shurshurik) [inline vpn-router-core/src/main/java/vpnrouter/core/infrastructure/shell/RealTunnelling.java:27]: shell commands seem so complicated to me 😱

File: vpn-router-core/src/main/java/vpnrouter/core/infrastructure/shell/RealTunnelling.java (hunk 1 above)
Comment (KirillSmirnow) [reply vpn-router-core/src/main/java/vpnrouter/core/infrastructure/shell/RealTunnelling.java:19]: https://habr.com/ru/articles/108690/
http://linux-ip.net/html/routing-tables.html
https://man7.org/linux/man-pages/man8/ip-rule.8.html

There are 4 default routing tables: unspec, local, main, and default. They have IDs 0, 255, 254, and 253 correspondingly. The default tables are declared in `/etc/iproute2/rt_tables`.

Therefore, we can choose any number (idk the range) except those default IDs for our custom routing table.

File: vpn-router-core/src/main/java/vpnrouter/core/infrastructure/shell/RealTunnelling.java (hunk 2 above)
Comment (KirillSmirnow) [reply vpn-router-core/src/main/java/vpnrouter/core/infrastructure/shell/RealTunnelling.java:27]: Then you can focus on Java code :)

File: vpn-router-core/src/main/java/vpnrouter/core/infrastructure/shell/RealTunnelling.java (hunk 1 above)
Comment (KirillSmirnow) [reply vpn-router-core/src/main/java/vpnrouter/core/infrastructure/shell/RealTunnelling.java:19]: image
<hr>
image

Comment (shurshurik) [reply vpn-router-core/src/main/java/vpnrouter/core/infrastructure/shell/RealTunnelling.java:19]: Thanks) After reading this links, I got some insights about it, but I'd like to discuss it one more time when we have a call)

**Review by shurshurik — APPROVED** [review]

**Pull request merged**